In [ ]:
False == (1 > 2), True == (2 > 1)

(True, True)

**False, True =Data values from the data type Boolean**.

In [ ]:
print((9 > 7) and (2 < 4))	# Both original expression is true.
print((8 == 8) or (6 != 6))	# One original expression is True.
print(not(3 <= 1))			# The original expression is False.

True
True
True


In [ ]:
s = 'Leaves'
# Using for loop
for letter in s:
  
    print(letter)
    # break the loop as soon it sees 'e'
    # or 's'
    if letter == 'e' or letter == 's':
        break
  
print("Out of for loop")
print()
  
  

L
e
Out of for loop



**Break statement in Python is used to bring the control out of the loop when some external condition is triggered.** 

In [ ]:
# loop from 1 to 10
for i in range(1, 11):
 
    # If i is equals to 6,
    # continue to next iteration
    # without printing
    if i == 6:
        continue
    else:
        # otherwise print the value
        # of i
        print(i, end=" ")

1 2 3 4 5 7 8 9 10 

**The continue statement rejects all the remaining statements in the current iteration of the loop and moves the control back to the top of the loop.**

**The continue statement can be used in both while and for loops.**



In [ ]:
num = 3.4
if num > 0:
    print("Positive number")
elif num == 0:
    print("Zero")
else:
    print("Negative number")

Positive number


**If the condition for if is False, it checks the condition of the next elif block and checks for else statements.**

In [ ]:
for x in range(6):
  print(x)

0
1
2
3
4
5


 **The for loop we can execute a set of statements, once for each item in a list, tuple, set etc.**

In [ ]:
i = 1
while i < 6:
  print(i)
  i += 1

1
2
3
4
5


 **The while loop we can execute a set of statements as long as a condition is true.**



In [ ]:
dict1 = {1: "one", 2: "two", 3: "three", 4: "four"}
 
print("one" in dict1)
print(3 in dict1)
print(5 in dict1)

False
True
False


**The in operator in Python checks whether a specified value is a constituent element of a sequence like string, array, list, or tuple etc.**

In [ ]:
 x = []
 y = x
 print(x is y)

True


**The operator x is y returns True if x and y are bound to the same object, otherwise False.**

In [ ]:
 value= None

if value is None: 
  print("None")
else:
  print("Not None")

None


**The None keyword is used to define a null variable or an object.**

In [ ]:
x = lambda a, b : a * b
print(x(5, 6))

30


**A lambda function is a small anonymous function ,lambda function can take any number of arguments, but can only have one expression.**

 ***A return statement is used to end the execution of the function call and “returns” the result (value of the expression following the return keyword) to the caller** 

In [ ]:
def my_function(x):
  return 5 * x

print(my_function(3))
print(my_function(5))
print(my_function(9))


15
25
45


***The bool() function allows you to evaluate any value, and give you True or False in return**



In [ ]:
print(bool("Hello"))
print(bool(15))

True
True


****The following operators which may be applied to numbers: +, -, * and **. 
**The division operator / for integers gives a floating-point real number (an object of type float). The exponentiation also returns a float when the power is negative**

In [ ]:
print(17 / 3)  # gives 5.66666666667
print(2 ** 4)  # gives 16
print(2 ** -2) # gives 0.25

5.666666666666667
16
0.25


***Strings in Python can be created using single quotes or double quotes or even triple quotes**. 

In [ ]:

# Python Program to Access
# characters of String
 
String1 = "GeeksForGeeks"
print("Initial String: ")
print(String1)
 
# Printing First character
print("\nFirst character of String is: ")
print(String1[0])
 
# Printing Last character
print("\nLast character of String is: ")
print(String1[-1])

Initial String: 
GeeksForGeeks

First character of String is: 
G

Last character of String is: 
s


***Lists are used to store multiple items in a single variable.**

In [ ]:
thislist = ["apple", "banana", "cherry"]
print(thislist)

['apple', 'banana', 'cherry']


***String, int and boolean data types and list can contain different data types:**

In [ ]:
list1 = ["apple", "banana", "cherry"]
list2 = [1, 5, 7, 9, 3]
list3 = [True, False, False]

print(list1)
print(list2)
print(list3)


['apple', 'banana', 'cherry']
[1, 5, 7, 9, 3]
[True, False, False]


***The remove() method removes the specified item.**(Remove "banana":)

In [ ]:
thislist = ["apple", "banana", "cherry"]
thislist.remove("banana")
print(thislist)


['apple', 'cherry']


***List comprehension offers a shorter syntax when you want to create a new list based on the values of an existing list.**

In [ ]:
fruits = ["apple", "banana", "cherry", "grape", "mango"]
newlist = []

for x in fruits:
  if "a" in x:
    newlist.append(x)

print(newlist)


['apple', 'banana', 'grape', 'mango']


***The index() method returns the position at the first occurrence of the specified value.**

In [ ]:
fruits = [4, 55, 64, 32, 16, 32]

x = fruits.index(32)


***List objects have a sort() method that will sort the list alphanumerically, ascending, by default:**

In [ ]:
thislist = ["orange", "mango", "kiwi", "pineapple", "banana"]
thislist.sort()
print(thislist)

['banana', 'kiwi', 'mango', 'orange', 'pineapple']


***The sort() method sorts the list ascending by default.**

In [ ]:
cars = ['Ford', 'BMW', 'Volvo']

cars.sort()

***To add an item to the end of the list, use the append() method:**

In [ ]:
thislist = ["apple", "banana", "cherry"]
thislist.append("orange")
print(thislist)


['apple', 'banana', 'cherry', 'orange']


***List items are indexed and you can access them by referring to the index number:t**

In [ ]:
thislist = ["apple", "banana", "cherry"]
print(thislist[1])

banana


***we can copy the list by using there are ways to make a copy, one way is to use the built-in List method copy()**.

In [ ]:
thislist = ["apple", "banana", "cherry"]
mylist = thislist.copy()
print(mylist)

['apple', 'banana', 'cherry']


***Dictionaries are used to store data values in key:value pairs**.

**A dictionary is a collection which is ordered*, changeable and do not allow duplicates.**

In [ ]:
thisdict = {
  "brand": "Ford",
  "model": "Mustang",
  "year": 1964
}
print(thisdict)


{'brand': 'Ford', 'model': 'Mustang', 'year': 1964}


***You can loop through a dictionary by using a for loop**.

**When looping through a dictionary, the return value are the keys of the dictionary, but there are methods to return the values as well.**

In [ ]:
for x in thisdict:
  print(x)

brand
model
year


*
**There are ways to make a copy, one way is to use the built-in Dictionary method copy().**

In [ ]:
thisdict = {
  "brand": "Ford",
  "model": "Mustang",
  "year": 1964
}
mydict = thisdict.copy()
print(mydict)

{'brand': 'Ford', 'model': 'Mustang', 'year': 1964}


***There are several methods to remove items from a dictionary.**

In [ ]:
thisdict = {
  "brand": "Ford",
  "model": "Mustang",
  "year": 1964
}
thisdict.pop("model")
print(thisdict)


{'brand': 'Ford', 'year': 1964}


***You can access the items of a dictionary by referring to its key name, inside square brackets:**

In [ ]:
thisdict = {
  "brand": "Ford",
  "model": "Mustang",
  "year": 1964
}
x = thisdict["model"]


***You can change the value of a specific item by referring to its key name**

In [ ]:
thisdict = {
  "brand": "Ford",
  "model": "Mustang",
  "year": 1964
}
thisdict["year"] = 2018


***Adding an item to the dictionary is done by using a new index key and assigning a value to it**

In [ ]:
thisdict = {
  "brand": "Ford",
  "model": "Mustang",
  "year": 1964
}
thisdict["color"] = "red"
print(thisdict)


{'brand': 'Ford', 'model': 'Mustang', 'year': 1964, 'color': 'red'}


***There are several methods to remove items from a dictionary**

In [ ]:
thisdict = {
  "brand": "Ford",
  "model": "Mustang",
  "year": 1964
}
thisdict.pop("model")
print(thisdict)


{'brand': 'Ford', 'year': 1964}


***A dictionary can contain dictionaries, this is called nested dictionaries**.



In [ ]:
myfamily = {
  "child1" : {
    "name" : "Emil",
    "year" : 2004
  },
  "child2" : {
    "name" : "Tobias",
    "year" : 2007
  },
  "child3" : {
    "name" : "Linus",
    "year" : 2011
  }
}
print(myfamily)


{'child1': {'name': 'Emil', 'year': 2004}, 'child2': {'name': 'Tobias', 'year': 2007}, 'child3': {'name': 'Linus', 'year': 2011}}


***A stack is a linear data structure where data is arranged objects on over another. It stores the data in LIFO (Last in First Out) manner.**

In [ ]:
my_stack =[]
  
# append() function is used to push   
# element in the my_stack   
my_stack.append('a')  
my_stack.append('b')  
my_stack.append('c')  
  
print('Initial my_stack:')  
print(my_stack)  
  
# pop() function is used to pop   
# element from my_stack in    
# LIFO order   
print('\nElements poped from my_stack:')  
print(my_stack.pop())  
print(my_stack.pop())  
print(my_stack.pop())  
  
print('\nmy_stack after elements are poped:')  
print(my_stack)   


Initial my_stack:
['a', 'b', 'c']

Elements poped from my_stack:
c
b
a

my_stack after elements are poped:
[]


***Sets are used to store multiple items in a single variable,A set is a collection which is unordered, unchangeable*and unindexed** .

In [ ]:
thisset = {"apple", "banana", "cherry"}
print(thisset)

{'cherry', 'apple', 'banana'}


***You cannot access items in a set by referring to an index or a key**

In [ ]:
thisset = {"apple", "banana", "cherry"}

for x in thisset:
  print(x)

cherry
apple
banana


***Once a set is created, you cannot change its items, but you can add new items.**

In [ ]:
thisset = {"apple", "banana", "cherry"}

thisset.add("orange")

print(thisset)


{'orange', 'cherry', 'apple', 'banana'}


***To remove an item in a set, use the remove(), or the discard() method**

In [ ]:
thisset = {"apple", "banana", "cherry"}

thisset.remove("banana")

print(thisset)

{'cherry', 'apple'}


***You can loop through the set items by using a for loop:**

In [ ]:
thisset = {"apple", "banana", "cherry"}

for x in thisset:
  print(x)

cherry
apple
banana


***Membership operators are used to test if a sequence is presented in an object:**

**Returns True if a sequence with the specified value is present in the object	x in y**(in operator)

In [ ]:
x = ["apple", "banana"]

print("banana" in x)

# returns True because a sequence with the value "banana" is in the list


True


***Returns True if a sequence with the specified value is not present in the object	x not in y**(not in operator

In [ ]:
x = ["apple", "banana"]

print("pineapple" not in x)

# returns True because a sequence with the value "pineapple" is not in the list

True


***A set comprehension is similar to a list comprehension but returns a set instead of a list,Without list comprehension you will have to write a for statement with a conditional test inside**:

In [ ]:
fruits = ["apple", "banana", "cherry", "kiwi", "mango"]
newlist = []

for x in fruits:
  if "a" in x:
    newlist.append(x)

print(newlist)

['apple', 'banana', 'mango']


***A Class is like an object constructor, or a "blueprint" for creating objects**.

**Create a class named MyClass, with a property named x**


In [ ]:
class MyClass:
  x = 5

print(MyClass)


<class '__main__.MyClass'>


**Now we can use the class named MyClass to create objects**.

**Create an object named p1, and print the value of xP**

In [ ]:
class MyClass:
  x = 5

p1 = MyClass()
print(p1.x)

5


***A class is a user-defined blueprint or prototype from which objects are created.**
**Instance attributes are those attributes that are not shared by objects. Every object has its own copy of the instance attribute**

In [ ]:
class Person:  
      
    # init method or constructor   
    def __init__(self, name):  
        self.name = name  
      
    # Sample Method   
    def say_hi(self):  
        print('Hello, my name is', self.name)  
      
p = Person('Nikhil')  
p.say_hi() 

Hello, my name is Nikhil


***self represents the instance of the class. By using the “self” keyword we can access the attributes and methods of the class in python. It binds the attributes with the given arguments**.
**The reason you need to use self. is because Python does not use the @ syntax to refer to instance attributes.**

In [ ]:


class car():
	
	# init method or constructor
	def __init__(self, model, color):
		self.model = model
		self.color = color
		
	def show(self):
		print("Model is", self.model )
		print("color is", self.color )
		
# both objects have different self which
# contain their attributes
audi = car("audi a4", "blue")
ferrari = car("ferrari 488", "green")

audi.show()	 # same output as car.show(audi)
ferrari.show() # same output as car.show(ferrari)

# Behind the scene, in every instance method
# call, python sends the instances also with
# that method call like car.show(audi)


Model is audi a4
color is blue
Model is ferrari 488
color is green


***map() function returns a map object(which is an iterator) of the results after applying the given function to each item of a given iterable (list, tuple etc.)**

**fun : It is a function to which map passes each element of given iterable**.

**iter : It is a iterable which is to be mapped**.

**The returned value from map() (map object) then can be passed to functions like list() (to create a list), set() (to create a set)  **

In [ ]:
# Python program to demonstrate working
# of map.
  
# Return double of n
def addition(n):
    return n + n
  
# We double all numbers using map()
numbers = (1, 2, 3, 4)
result = map(addition, numbers)
print(list(result))

[2, 4, 6, 8]


 **Add two lists using map and lambda**

In [ ]:
numbers1 = [1, 2, 3]
numbers2 = [4, 5, 6]

result = map(lambda x, y: x + y, numbers1, numbers2)
print(list(result))

[5, 7, 9]


***The map in Python takes a function and an iterable/iterables. It loops over each item of an iterable and applies the transformation function to it. Then, it returns a map object that stores the value of the transformed item**.

***Using For Loop**

In [ ]:
num = [3, 5, 7, 11, 13]

mul = []

for n in num:

    mul.append(n ** 2)

print (mul)



[9, 25, 49, 121, 169]


***The join() string method returns a string by joining all the elements of an iterable (list, string, tuple), separated by a string separator.t**

In [ ]:
text = ['Python', 'is', 'a', 'fun', 'programming', 'language']

# join elements of text with space
print(' '.join(text))


Python is a fun programming language


***The string ljust() method returns a left-justified string of a given minimum width.**

In [ ]:
string = 'cat'
width = 5

# print left justified string
print(string.ljust(width))

cat  


***The string rjust() method returns a right-justified string of a given minimum width**

In [ ]:
string = 'cat'
width = 5

# print right justified string
print(string.rjust(width))

  cat


***The center() method returns a string which is padded with the specified character.**

In [ ]:
string = "Python is awesome"

new_string = string.center(24)

print("Centered String: ", new_string)


Centered String:     Python is awesome    


***The index() method returns the index of a substring inside the string (if found). If the substring is not found, it raises an exception.**

In [ ]:
text = 'Python is fun'

# find the index of is
result = text.index('is')
print(result)



7


 ***The filter() method filters the given sequence with the help of a function that tests each element in the sequence to be true or not.**

In [ ]:
# function that filters vowels
def fun(variable):
	letters = ['a', 'e', 'i', 'o', 'u']
	if (variable in letters):
		return True
	else:
		return False


# sequence
sequence = ['g', 'e', 'e', 'j', 'k', 's', 'p', 'r']

# using filter function
filtered = filter(fun, sequence)

print('The filtered letters are:')
for s in filtered:
	print(s)


The filtered letters are:
e
e


***The strip() method removes any leading (spaces at the beginning) and trailing (spaces at the end) characters (space is the default leading character to remove).**

**Remove the leading and trailing characters**

In [ ]:
txt = ",,,,,rrttgg.....banana....rrr"

x = txt.strip(",.grt")

print(x)


banana


***Python sorted() function returns a sorted list from the iterable object**.

**Sorted() sorts any sequence (list, tuple) and always returns a list with the elements in a sorted manner, without modifying the original sequence.**

**sorted takes three parameters from which two are optional.
Iterable : sequence (list, tuple, string) or collection (dictionary, set, frozenset) or any other iterator that needs to be sorted.
Key(optional) : A function that would server as a key or a basis of sort comparison.
Reverse(optional) : If set true, then the iterable would be sorted in reverse (descending) order, by default it is set as false.

In [ ]:
x = [2, 8, 1, 4, 6, 3, 7]

print("Sorted List returned :"),
print(sorted(x))

print("\nReverse sort :"),
print(sorted(x, reverse=True))

print("\nOriginal list not modified :"),
print(x)


Sorted List returned :
[1, 2, 3, 4, 6, 7, 8]

Reverse sort :
[8, 7, 6, 4, 3, 2, 1]

Original list not modified :
[2, 8, 1, 4, 6, 3, 7]


 ***Python sorted reverse**

In [ ]:
# Python3 code to demonstrate
# Reverse Sort a String
# using join() + sorted() + reverse

# initializing string
test_string = "geekforgeeks"

# printing original string
print("The original string : " + str(test_string))

# using join() + sorted() + reverse
# Sorting a string
res = ''.join(sorted(test_string, reverse = True))
	
# print result
print("String after reverse sorting : " + str(res))


The original string : geekforgeeks
String after reverse sorting : srokkggfeeee


***Python sorted() lambda**

In [ ]:
# Python3 code to demonstrate
# Reverse Sort a String
# using join() + sorted() + reverse

# initializing string
test_string = "geekforgeeks"

# printing original string
print("The original string : " + str(test_string))

# using join() + sorted() + reverse
# Sorting a string
res = ''.join(sorted(test_string, reverse = True))
	
# print result
print("String after reverse sorting : " + str(res))


The original string : geekforgeeks
String after reverse sorting : srokkggfeeee


***Sorting means rearranging a given sequence of elements according to a comparison operator on the elements. The comparison operator is used to decide the new order of the element in the respective data structure.**

***Python sorted() key**

In [ ]:
# Python program to demonstrate
# sorted()


L = [1, 2, 3, 4, 5]

print("Sorted list:")
print(sorted(L))

print("\nReverse sorted list:")
print(sorted(L, reverse = True))

print("\nOriginal list after sorting:")
print(L)


Sorted list:
[1, 2, 3, 4, 5]

Reverse sorted list:
[5, 4, 3, 2, 1]

Original list after sorting:
[1, 2, 3, 4, 5]


***Using key parameter**

In [ ]:
# Python program to demonstrate sorting by user's
# choice

# function to return the second element of the
# two elements passed as the parameter
def sortSecond(val):
	return val[1]

# list1 to demonstrate the use of sorting
# using using second key
list1 = [(1, 2), (3, 3), (1, 1)]

# sorts the array in ascending according to
# second element
list1.sort(key = sortSecond)
print(list1)

# sorts the array in descending according to
# second element
list1.sort(key = sortSecond, reverse = True)
print(list1)


[(1, 1), (1, 2), (3, 3)]
[(3, 3), (1, 2), (1, 1)]


***Both list.sort() and sorted() have a key parameter to specify a function (or other callable) to be called on each list element prior to making comparisons.**

In [ ]:
sorted("This is a test string from Andrew".split(), key=str.lower)
['a', 'Andrew', 'from', 'is', 'string', 'test', 'This']

['a', 'Andrew', 'from', 'is', 'string', 'test', 'This']

***The Python help function is used to display the documentation of modules, functions, classes, keywords, etc**.

**If the help function is passed without an argument, then the interactive help utility starts up on the console.**

In [ ]:
help(print)


Help on built-in function print in module builtins:

print(...)
    print(value, ..., sep=' ', end='\n', file=sys.stdout, flush=False)
    
    Prints the values to a stream, or to sys.stdout by default.
    Optional keyword arguments:
    file:  a file-like object (stream); defaults to the current sys.stdout.
    sep:   string inserted between values, default a space.
    end:   string appended after the last value, default a newline.
    flush: whether to forcibly flush the stream.



***Help function output can also be defined for user-defined functions and classes. The docstring(documentation string) is used for documentation. It is nested inside triple quotes and is the first statement within a class or function or a module.**

In [ ]:
class Helper:
	def __init__(self):
		'''The helper class is initialized'''

	def print_help(self):
		'''Returns the help description'''
		print('helper description')


help(Helper)
help(Helper.print_help)


Help on class Helper in module __main__:

class Helper(builtins.object)
 |  Methods defined here:
 |  
 |  __init__(self)
 |      The helper class is initialized
 |  
 |  print_help(self)
 |      Returns the help description
 |  
 |  ----------------------------------------------------------------------
 |  Data descriptors defined here:
 |  
 |  __dict__
 |      dictionary for instance variables (if defined)
 |  
 |  __weakref__
 |      list of weak references to the object (if defined)

Help on function print_help in module __main__:

print_help(self)
    Returns the help description



***Python help() function docstring**

**The docstrings are declared using ”’triple single quotes”’ or “””triple double quotes””” just below the class, method or function declaration. All functions should have a docstring**.

**Accessing Docstrings: The docstrings can be accessed using the __doc__ method of the object or using the help function.**

In [ ]:
def my_function():
	'''Demonstrates triple double quotes
	docstrings and does nothing really.'''

	return None

print("Using __doc__:")
print(my_function.__doc__)

print("Using help:")
help(my_function)


Using __doc__:
Demonstrates triple double quotes
	docstrings and does nothing really.
Using help:
Help on function my_function in module __main__:

my_function()
    Demonstrates triple double quotes
    docstrings and does nothing really.



***The zip() function returns a zip object, which is an iterator of tuples where the first item in each passed iterator is paired together, and then the second item in each passed iterator are paired together etc.

If the passed iterators have different lengths, the iterator with the least items decides the length of the new iterator.**

If one tuple contains more items, these items are ignored

In [ ]:
a = ("yamu", "sri","suji")
b = ("aarthi", "yamini", "pavi", "Vicky")

x = zip(a, b)
print(tuple(x))

(('yamu', 'aarthi'), ('sri', 'yamini'), ('suji', 'pavi'))


***The zipping techniques, that is assigning a key-value or pairing from two different lists has been covered in many articles before, sometimes we have specific utility to perform the reverse task. This task can be achieved by various methods.**

In [ ]:
# Python3 code to demonstrate
# Unzip a list of tuples
# using list comprehension

# initializing list of tuples
test_list = [('Akshat', 1), ('Bro', 2), ('is', 3), ('Placed', 4)]

# Printing original list
print ("Original list is : " + str(test_list))

# using list comprehension to
# perform Unzipping
res = [[ i for i, j in test_list ],
	[ j for i, j in test_list ]]
	
# Printing modified list
print ("Modified list is : " + str(res))


Original list is : [('Akshat', 1), ('Bro', 2), ('is', 3), ('Placed', 4)]
Modified list is : [['Akshat', 'Bro', 'is', 'Placed'], [1, 2, 3, 4]]


***Using zip() and * operator**

**Mostly used method to perform unzip and most Pythonic and recommended as well. This method is generally used to perform this task by programmers all over. * operator unzips the tuples into independent lists**.

In [ ]:
# Python3 code to demonstrate
# Unzip a list of tuples
# using zip() and * operator

# initializing list of tuples
test_list = [('Akshat', 1), ('Bro', 2), ('is', 3), ('Placed', 4)]

# Printing original list
print ("Original list is : " + str(test_list))

# using zip() and * operator to
# perform Unzipping
res = list(zip(*test_list))
	
# Printing modified list
print ("Modified list is : " + str(res))


Original list is : [('Akshat', 1), ('Bro', 2), ('is', 3), ('Placed', 4)]
Modified list is : [('Akshat', 'Bro', 'is', 'Placed'), (1, 2, 3, 4)]


***Enumerate() in Python**

***Enumerate() method adds a counter to an iterable and returns it in a form of enumerating object. This enumerated object can then be used directly for loops or converted into a list of tuples using the list() method.**

In [ ]:
# python
# Python program to illustrate
# enumerate function
l1 = ["eat", "sleep", "repeat"]
s1 = "geek"

# creating enumerate objects
obj1 = enumerate(l1)
obj2 = enumerate(s1)

print ("Return type:", type(obj1))
print (list(enumerate(l1)))

# changing start index to 2 from 0
print (list(enumerate(s1, 2)))


Return type: <class 'enumerate'>
[(0, 'eat'), (1, 'sleep'), (2, 'repeat')]
[(2, 'g'), (3, 'e'), (4, 'e'), (5, 'k')]


***Using Enumerate object in loops:**

In [ ]:
# Python program to illustrate
# enumerate function in loops
l1 = ["eat", "sleep", "repeat"]

# printing the tuples in object directly
for ele in enumerate(l1):
	print (ele)

# changing index and printing separately
for count, ele in enumerate(l1, 100):
	print (count, ele)

# getting desired output from tuple
for count, ele in enumerate(l1):
	print(count)
	print(ele)


(0, 'eat')
(1, 'sleep')
(2, 'repeat')
100 eat
101 sleep
102 repeat
0
eat
1
sleep
2
repeat


***Swap Two Variables**

**Given two variables x and y, write a Python program to swap their values**

     swap two numbers ,first number 10 and second number 50.

***Using Naive approach.

     The most naive approach is to store the value of one variable(say x) in a temporary variable,
      then assigning the variable x with the value of variable y.
       Finally, assign the variable y with the value of the temporary variable.

In [ ]:
# Python program to demonstrate
# swapping of two variables

x = 10
y = 50

# Swapping of two variables
# Using third variable
temp = x
x = y
y = temp

print("Value of x:", x)
print("Value of y:", y)


Value of x: 50
Value of y: 10


***Using comma operator**

  Using the comma operator the value of variables can be swapped without using a third variable.

In [ ]:
# Python program to demonstrate
# swapping of two variables


x = 10
y = 50

# Swapping of two variables
# without using third variable
x, y = y, x

print("Value of x:", x)
print("Value of y:", y)


Value of x: 50
Value of y: 10


***Using XOR**

  The bitwise XOR operator can be used to swap two variables. The XOR of two numbers x and y returns a number which has all the bits as 1 wherever bits of x and y differ. For example XOR of 10 (In Binary 1010) and 5 (In Binary 0101) is 1111 and XOR of 7 (0111) and 5 (0101) is (0010).
 

In [ ]:
# Python program to demonstrate
# Swapping of two variables

x = 10
y = 50

# Swapping using xor
x = x ^ y
y = x ^ y
x = x ^ y

print("Value of x:", x)
print("Value of y:", y)


Value of x: 50
Value of y: 10


***Using arithmetic operators we can perform swapping in two ways.**

      Using addition and subtraction operator :
The idea is to get sum in one of the two given numbers. The numbers can then be swapped using the sum and subtraction from sum.
 
 

In [ ]:
# Python program to demonstrate
# swapping of two variables

x = 10
y = 50

# Swapping of two variables
# using arithmetic operations
x = x + y
y = x - y
x = x - y

print("Value of x:", x)
print("Value of y:", y)


Value of x: 50
Value of y: 10


***Using multiplication and division operator :**

  The idea is to get multiplication of the two given numbers. The numbers can then be calculated by using the division.
    

In [ ]:
# Python program to demonstrate
# swapping of two variables

x = 10
y = 50

# Swapping of two numbers
# Using multiplication operator

x = x * y
y = x / y
x = x / y

print("Value of x : ", x)
print("Value of y : ", y)


Value of x :  50.0
Value of y :  10.0


***using Bitwise addition and subtraction for swapping**
 

In [ ]:
#Python program to demonstrate
#swapping of two numbers
a = 5
b = 1
a = (a & b) + (a | b)
b = a + (~b) + 1
a = a + (~b) + 1
print("a after swapping: ", a)
print("b after swapping: ", b)


a after swapping:  1
b after swapping:  5


***We can use * to unpack the list so that all elements of it can be passed as different parameters.**

In [ ]:
# A sample function that takes 4 arguments
# and prints the,
def fun(a, b, c, d):
	print(a, b, c, d)

# Driver Code
my_list = [1, 2, 3, 4]

# Unpacking list into four arguments
fun(*my_list)


1 2 3 4


***PACKING**
  **When we don’t know how many arguments need to be passed to a python function, we can use Packing to pack all arguments in a tuple. *

In [ ]:
# A Python program to demonstrate use
# of packing

# This function uses packing to sum
# unknown number of arguments
def mySum(*args):
	return sum(args)

# Driver code
print(mySum(1, 2, 3, 4, 5))
print(mySum(10, 20))


15
30


***Packing and Unpacking**
  Below is an example that shows both packing and unpacking. **

In [ ]:
# A Python program to demonstrate both packing and
# unpacking.

# A sample python function that takes three arguments
# and prints them
def fun1(a, b, c):
	print(a, b, c)

# Another sample function.
# This is an example of PACKING. All arguments passed
# to fun2 are packed into tuple *args.
def fun2(*args):

	# Convert args tuple to a list so we can modify it
	args = list(args)

	# Modifying args
	args[0] = 'Geeksforgeeks'
	args[1] = 'awesome'

	# UNPACKING args and calling fun1()
	fun1(*args)

# Driver code
fun2('Hello', 'beautiful', 'world!')


Geeksforgeeks awesome world!


****is used for dictionaries**

In [ ]:
# A sample program to demonstrate unpacking of
# dictionary items using **
def fun(a, b, c):
	print(a, b, c)

# A call with unpacking of dictionary
d = {'a':2, 'b':4, 'c':10}
fun(**d)


2 4 10


***This designates a subexpression that will be assigned a list of all items from the iterable being unpacked that are not assigned to any of the mandatory expressions, or an empty list if there are no such items.**

In [1]:
seq = [1, 2, 3, 4]
a, b, c, d, *e = seq
print(a, b, c, d, e)

1 2 3 4 []


****We can use * to unpack the list so that all elements of it can be passed as different parameters.**

In [2]:
# A sample function that takes 4 arguments
# and prints the,
def fun(a, b, c, d):
	print(a, b, c, d)

# Driver Code
my_list = [1, 2, 3, 4]

# Unpacking list into four arguments
fun(*my_list)


1 2 3 4


***merge two dictionaries**

   ***There are various ways in which Dictionaries can be merged by the use of various functions and constructors in Python. In this article, we will discuss a few ways of merging dictionaries**. 

   **Using the method update() **

   **By using the method update() in Python, one list can be merged into another. But in this, the second list is merged into the first list and no new list is created,It returns None**
   

In [3]:
# Python code to merge dict using update() method
def Merge(dict1, dict2):
	return(dict2.update(dict1))
	
# Driver code
dict1 = {'a': 10, 'b': 8}
dict2 = {'d': 6, 'c': 4}

# This return None
print(Merge(dict1, dict2))

# changes made in dict2
print(dict2)


None
{'d': 6, 'c': 4, 'a': 10, 'b': 8}


Using ** in Python

In [4]:
# Python code to merge dict using a single
# expression
def Merge(dict1, dict2):
	res = {**dict1, **dict2}
	return res
	
# Driver code
dict1 = {'a': 10, 'b': 8}
dict2 = {'d': 6, 'c': 4}
dict3 = Merge(dict1, dict2)
print(dict3)


{'a': 10, 'b': 8, 'd': 6, 'c': 4}
